$MirOS: ports/sysutils/ntfsprogs/patches/patch-libntfs_attrib_c,v 1.1.7.1 2005/03/18 15:52:18 tg Exp $
$NetBSD: patch-ac,v 1.1.1.1 2004/10/05 12:39:38 agc Exp $
--- libntfs/attrib.c.orig	Mon Mar 27 20:01:13 2006
+++ libntfs/attrib.c	Sun Apr 29 22:26:06 2007
@@ -763,7 +763,7 @@ map_rl:
  */
 s64 ntfs_attr_pread(ntfs_attr *na, const s64 pos, s64 count, void *b)
 {
-	s64 br, to_read, ofs, total, total2;
+	s64 br, to_read, ofs, total, total2, origcount;
 	ntfs_volume *vol;
 	runlist_element *rl;
 
@@ -795,6 +795,7 @@ s64 ntfs_attr_pread(ntfs_attr *na, const
 		ntfs_inode_update_atime(na->ni);
 	if (!count)
 		return 0;
+	origcount = count;
 	/* Truncate reads beyond end of attribute. */
 	if (pos + count > na->data_size) {
 		if (pos >= na->data_size)
@@ -885,13 +886,19 @@ res_err_out:
 			continue;
 		}
 		/* It is a real lcn, read it into @dst. */
-		to_read = min(count, (rl->length << vol->cluster_size_bits) -
+		to_read = min(origcount, (rl->length << vol->cluster_size_bits) -
 				ofs);
 retry:
 		ntfs_log_trace("Reading 0x%llx bytes from vcn 0x%llx, lcn 0x%llx, "
 				"ofs 0x%llx.\n", to_read, rl->vcn, rl->lcn, ofs);
 		br = ntfs_pread(vol->dev, (rl->lcn << vol->cluster_size_bits) +
 				ofs, to_read, b);
+		/* Zero any overage that we may have read */
+		if (br > count) {
+			memset(b + count, 0, br - count);
+			total += count;
+			return total + total2;
+		}
 		/* If everything ok, update progress counters and continue. */
 		if (br > 0) {
 			total += br;
